src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java

```diff
@@ -0,0 +1,274 @@
+    }
+
+    /**
+     * @return true if there are 0 elements in the pileup, false otherwise
+     */
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    /**
+     * @return the location of this pileup.
+     */
+    public Locatable getLocation() {
+        return loc;
+    }
+
+    /**
+     * Get counts of A, C, G, T in order, which returns a int[4] vector with counts according
+     * to BaseUtils.simpleBaseToBaseIndex for each base.
+     * Deletions are not counted.
+     */
+    public int[] getBaseCounts() {
+        final int[] counts = new int[4];
+
+        for (final PileupElement pile : this) {
+            // skip deletion sites
+            if (!pile.isDeletion()) {
+                final int index = BaseUtils.simpleBaseToBaseIndex(pile.getBase());
+                if (index != -1) {
+                    counts[index]++;
+                }
+            }
+        }
+
+        return counts;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%s %s %s %s",
+                loc.getContig(),
+                loc.getStart(),
+                new String(getBases()),
+                getQualsString());
+    }
+
+    /**
+     * Returns a list of the reads in this pileup. Note this call costs O(n) and allocates fresh lists each time
+     */
+    public List<GATKRead> getReads() {
+        return pileupElements.stream().map(pe -> pe.getRead()).collect(Collectors.toList());
+    }
+
+    /**
+     * Returns the number of elements that satisfy the predicate.
+     */
+    public int getNumberOfElements(final Predicate<PileupElement> peFilter){
+        Utils.nonNull(peFilter);
+        //Note: pileups are small so int is fine.
+        return (int)pileupElements.stream().filter(peFilter).count();
+    }
+
+    /**
+     * Returns a list of the offsets in this pileup.
+     * Note: this call costs O(n) and allocates fresh lists each time
+     */
+    public List<Integer> getOffsets() {
+        return pileupElements.stream().map(pe -> pe.getOffset()).collect(Collectors.toList());
+    }
+
+    //Extracts an int array by mapping each element in the pileup to int.
+    private int[] extractIntArray(final ToIntFunction<PileupElement> map){
+        return pileupElements.stream().mapToInt(map).toArray();
+    }
+
+    /**
+     * Returns an array of the bases in this pileup.
+     * Note: this call costs O(n) and allocates fresh array each time
+     */
+    public byte[] getBases() {
+        return toByteArray(extractIntArray(pe -> pe.getBase()));
+    }
+
+    /**
+     * Returns an array of the quals in this pileup.
+     * Note: this call costs O(n) and allocates fresh array each time
+     */
+    public byte[] getBaseQuals() {
+        return toByteArray(extractIntArray(pe -> pe.getQual()));
+    }
+
+    //Converts array of ints to array of bytes by hard casting (loses precision if ints are large).
+    private byte[] toByteArray(final int[] ints) {
```

Comment:
actually, deleted this method after addressing the byte->int->byte comments below. it was no longer used.